Add partially ordered sets

benlorenz opened the pull request:
Extracted from #3819, but might need some things from there to work.

Todo:
- ~~more tests (that are not doctests)~~
- ~~more docs~~

Closes https://github.com/oscar-system/Oscar.jl/pull/3610

I also added indegree and outdegree for graphs, fixes #4440.

benlorenz committed:
WIP: posets

benlorenz committed:
bibrefs

benlorenz committed:
posets: exports, docs, improvements

benlorenz commented:
Preview including doctests as examples: 
https://docs.oscar-system.org/previews/PR3928/Combinatorics/posets/

benlorenz committed:
fix some whitespace

micjoswig commented:
Looks good to me.

lgoettgens commented on src/Combinatorics/PartiallyOrderedSet/functions.jl:
````diff
@@ -0,0 +1,625 @@
+
+Return the rank of a partially ordered set, i.e., the rank of a maximal element.
+
+```jldoctest
````
```suggestion
# Examples
```jldoctest
```
our docstrings usually have a header between description and doctest (same applies to most other docstrings in this PR)

lgoettgens commented on src/Combinatorics/PartiallyOrderedSet/functions.jl:
```diff
@@ -0,0 +1,625 @@
+@doc raw"""
+    maximal_ranked_poset(v::AbstractVector{Int})
+
+Maximal ranked partially ordered set with number of nodes per rank given in
```
This sentence misses a verb

lgoettgens reviewed (commented):
I don't have any particular opinion on posets, but I found some minor stylistic things with the docstrings.
Iirc @felix-roehrich had some application of posets, so it would be good to hear from them if this would work

micjoswig committed:
poset docs

felix-roehrich commented:
For my use case I need a possibility to associate weights to the covering relations and I would need to experiment to see if this works well with the features here, but I currently don't have the time, so I cannot provide any useful comments.

However, I want to remind that there is already some functionality at src/NumberTheory/GaloisGrp/POSet.jl by @fieker. So maybe he has some opinions on this.

micjoswig commented:
> However, I want to remind that there is already some functionality at src/NumberTheory/GaloisGrp/POSet.jl by @fieker. So maybe he has some opinions on this.

The connection to POSet.jl was discussed in several zoom meetings.  The key difference is that, for the Galois groups, the poset is constructed dynamically.  This is not what we want here.   There is no good way of capturing both versions of posets in one framwork; the use cases are too varied.  Here we need posets as combinatorial objects, which are used for the construction of other objects, combinatorial or algebraic.  Since the latter constructions are typically large and costly, we cannot afford to waste running time for a design which is too abtract/slow.

The conclusion is to keep both poset implementations, at least for now.  One is exposed in the docs (this one), the other one is like an internal data structure.

micjoswig commented:
> For my use case I need a possibility to associate weights to the covering relations and I would need to experiment to see if this works well with the features here, but I currently don't have the time, so I cannot provide any useful comments.

It is not a good idea to overload the design of a general type with too many specifics.  There is a multitude of possibilities to keep information of your kind extra.

Just to give one example: `partially_ordered_set(covrels::Matrix{Int})` can generate a poset from a matrix. So you can store the pair of your weight matrix and the poset.  If you want to make it look more elegant, create an extra type.

@benlorenz pointed out that OSCAR already supports (directed and undirected) graphs with arbitrary edge weights/labels.  This could also be used for that purpose.

micjoswig committed:
poset doc edits; renamed face_lattice -> face_poset

fingolfin commented on src/Combinatorics/PartiallyOrderedSet/functions.jl:
````diff
@@ -0,0 +1,630 @@
+Partially ordered set of rank 3 on 6 elements
+```
+
+The construction only distinguishes zero coefficients (to indicate the absence) and nonzero coefficients (to indicate the presence of a covering relation).
````
Perhaps move this before the example?

fingolfin reviewed (approved):
I have no objections.

benlorenz committed:
example header

lgoettgens commented:
@benlorenz just said in triage that he is working on some small change, so converting this to a draft until this is in.

benlorenz committed:
Graphs: add indegree and outdegree

benlorenz committed:
Posets: docs improvements, allow artificial bottom node

benlorenz committed:
posets: improve support for artificial bottom/top

benlorenz committed:
groebnerwalk: add teardown for verbosity

to fix re-running doctests

benlorenz committed:
posets: construct from adjacency, fixes for elements

benlorenz committed:
posets: formatting and more matrixtypes

benlorenz committed:
test for in/out degree

benlorenz committed:
posets: small fixes, allow max chains as elements

benlorenz committed:
posets: tests

benlorenz committed:
docs meta block

benlorenz committed:
graph: add refs for indegree and outdegree

benlorenz committed:
fix docs

benlorenz commented:
I added proper tests, improved some input types (e.g. allowing sparse matrices), and added support for an artificial bottom node. This should be done from my point of view and can hopefully be merged in time for #4877.

lgoettgens commented:
@micjoswig could you please give this a review/approval with the latest changes (artificial bottom node etc.) so that we can merge this in time for Oscar 1.4?

micjoswig reviewed (approved):
OK.

lgoettgens commented:
I don't know if it is something flaky with the filesystem again, but we are experiencing issues in this new testset in the downstream tests in Nemo, see https://github.com/Nemocas/Nemo.jl/actions/runs/15140439210/job/42562858114?pr=2079#step:7:693.
Full error log:
```julia
Posets: Error During Test at /home/runner/work/Nemo.jl/Nemo.jl/oscar-dev/Oscar/test/Combinatorics/Posets.jl:1
  Got exception outside of a @test
  Can't load shared module /tmp/poly3824N1_aaaa0063.so: /opt/hostedtoolcache/julia/1.6.7/x64/bin/../lib/julia/libstdc++.so.6: version `GLIBCXX_3.4.32' not found (required by /tmp/poly3824N1_aaaa0063.so) at /home/runner/.julia/scratchspaces/d720cf60-89b5-51f5-aff5-213f193123e7/polymake_16401842906677952067_1.6_depstree_v2/share/polymake/perllib/Polymake/Core/CPlusPlus.pm line 1785.
  
  
  Stacktrace:
    [1] internal_call_function(arg1::String, arg2::CxxWrap.StdLib.StdVectorAllocated{CxxWrap.StdLib.StdString}, arg3::Vector{Any})
      @ Polymake.LibPolymake ~/.julia/packages/CxxWrap/I0P2E/src/CxxWrap.jl:676
    [2] #44
      @ ~/.julia/packages/Polymake/JoudB/src/call_function.jl:24 [inlined]
    [3] disable_sigint(f::Polymake.var"#44#45"{Symbol, CxxWrap.StdLib.StdVectorAllocated{CxxWrap.StdLib.StdString}, Vector{Any}, String})
      @ Base ./c.jl:458
    [4] call_function(::Type{Polymake.LibPolymake.PropertyValue}, ::Symbol, ::Symbol, ::Polymake.LibPolymake.GraphAllocated{Directed}, ::Vararg{Any, N} where N; template_parameters::Vector{String}, calltype::Symbol, kwargs::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
      @ Polymake ~/.julia/packages/Polymake/JoudB/src/call_function.jl:17
    [5] call_function(::Symbol, ::Symbol, ::Polymake.LibPolymake.GraphAllocated{Directed}, ::Vararg{Any, N} where N; template_parameters::Vector{String}, kwargs::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
      @ Polymake ~/.julia/packages/Polymake/JoudB/src/call_function.jl:31
    [6] call_function(::Symbol, ::Symbol, ::Polymake.LibPolymake.GraphAllocated{Directed}, ::Vector{Int64})
      @ Polymake ~/.julia/packages/Polymake/JoudB/src/call_function.jl:31
    [7] partially_ordered_set(gin::Graph{Directed})
      @ Oscar ~/work/Nemo.jl/Nemo.jl/oscar-dev/Oscar/src/Combinatorics/PartiallyOrderedSet/functions.jl:161
    [8] partially_ordered_set_from_adjacency
      @ ~/work/Nemo.jl/Nemo.jl/oscar-dev/Oscar/src/Combinatorics/PartiallyOrderedSet/functions.jl:55 [inlined]
    [9] partially_ordered_set(covrels::Polymake.LibPolymake.IncidenceMatrixAllocated{Polymake.LibPolymake.NonSymmetric})
      @ Oscar ~/work/Nemo.jl/Nemo.jl/oscar-dev/Oscar/src/Combinatorics/PartiallyOrderedSet/functions.jl:38
   [10] macro expansion
      @ ~/work/Nemo.jl/Nemo.jl/oscar-dev/Oscar/test/Combinatorics/Posets.jl:23 [inlined]
   [11] macro expansion
      @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
   [12] top-level scope
      @ ~/work/Nemo.jl/Nemo.jl/oscar-dev/Oscar/test/Combinatorics/Posets.jl:3
   [13] include
      @ ./Base.jl:385 [inlined]
   [14] macro expansion
      @ ./timing.jl:368 [inlined]
   [15] _timed_include(str::String, mod::Module; use_ctime::Bool)
      @ Oscar ~/work/Nemo.jl/Nemo.jl/oscar-dev/Oscar/src/utils/tests.jl:5
   [16] (::Oscar.var"#40#41"{Bool, Bool, Bool, Vector{Any}})()
      @ Oscar ~/work/Nemo.jl/Nemo.jl/oscar-dev/Oscar/src/utils/tests.jl:259
   [17] with_unicode(f::Oscar.var"#40#41"{Bool, Bool, Bool, Vector{Any}}, allowed::Bool)
      @ AbstractAlgebra.PrettyPrinting ~/.julia/packages/AbstractAlgebra/QGEmZ/src/PrettyPrinting.jl:1852
   [18] #test_module#39
      @ ~/work/Nemo.jl/Nemo.jl/oscar-dev/Oscar/src/utils/tests.jl:207 [inlined]
   [19] #14
      @ ~/work/Nemo.jl/Nemo.jl/oscar-dev/Oscar/test/runtests.jl:174 [inlined]
   [20] (::Base.var"#834#839"{var"#14#15"})(r::Base.RefValue{Any}, args::Tuple{String})
      @ Base ./asyncmap.jl:100
   [21] macro expansion
      @ ./asyncmap.jl:234 [inlined]
   [22] (::Base.var"#850#851"{Base.var"#834#839"{var"#14#15"}, Channel{Any}, Nothing})()
      @ Base ./task.jl:417
```

benlorenz commented:
This is a known issue when using an old julia version (with an old bundled C++ library) but with a recent gcc version in the system. We have workarounds in various places, e.g.:
https://github.com/oscar-system/Oscar.jl/blob/f2170d70ea74ee3cc7180af0e407e282a4dfeb09/.github/workflows/CI.yml#L102-L104
This is also listed in our FAQ: https://docs.oscar-system.org/dev/General/faq/

This only triggers when some polymake code triggers a new automatic compilation, I usually try to avoid this by providing as many pre-compiled instances as possible but this cannot be avoided in general.
